clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java

```diff
@@ -912,9 +910,15 @@ private boolean isProtocolTypeInconsistent(String protocolType) {
         return protocolType != null && !protocolType.equals(protocolType());
     }
 
-    private boolean isProtocolNameInconsistent(String protocolName) {
-        return protocolName != null && generation() != Generation.NO_GENERATION
-                   && !protocolName.equals(generation().protocolName);
+    private synchronized boolean isProtocolNameInconsistent(ApiKeys key, String protocolName) {
+        final boolean protocolNameInconsistent = protocolName != null &&
+            generation() != Generation.NO_GENERATION && !protocolName.equals(generation().protocolName);
+
+        if (protocolNameInconsistent) {
```

Comment:
@dajac just to clarify, are you concerning that the `generation()` may change between the check and the error-log? If yes maybe we do not need to synchronize the whole function, instead we just get a reference of the returned `generation()` call and use that in the error-log, since the generation object is immutable.

Reply:
@guozhangwang Yes, keeping the reference is fine. I was concerned by the check itself because we were calling `generation()` twice in the previous implementation thus we could get two different instances.
```
generation() != Generation.NO_GENERATION && !protocolName.equals(generation().protocolName)
```
I haven't thought about the error-log but it is also a good point.